plugins/locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/mylocationlayer/MyLocationLayerPlugin.java

```diff
@@ -0,0 +1,525 @@
+package com.mapbox.mapboxsdk.plugins.mylocationlayer;
+
+import android.animation.TypeEvaluator;
+import android.animation.ValueAnimator;
+import android.location.Location;
+import android.os.SystemClock;
+import android.support.annotation.NonNull;
+import android.view.animation.AccelerateDecelerateInterpolator;
+import android.view.animation.LinearInterpolator;
+
+import com.mapbox.mapboxsdk.location.LocationSource;
+import com.mapbox.mapboxsdk.maps.MapView;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+import com.mapbox.mapboxsdk.style.layers.Layer;
+import com.mapbox.mapboxsdk.style.layers.Property;
+import com.mapbox.mapboxsdk.style.layers.PropertyFactory;
+import com.mapbox.mapboxsdk.style.sources.GeoJsonSource;
+import com.mapbox.services.android.telemetry.location.LocationEngine;
+import com.mapbox.services.android.telemetry.location.LocationEngineListener;
+import com.mapbox.services.android.telemetry.location.LocationEnginePriority;
+import com.mapbox.services.api.utils.turf.TurfConstants;
+import com.mapbox.services.commons.geojson.Feature;
+import com.mapbox.services.commons.geojson.FeatureCollection;
+import com.mapbox.services.commons.geojson.Point;
+import com.mapbox.services.commons.geojson.Polygon;
+import com.mapbox.services.commons.models.Position;
+
+/**
+ * The My Location layer plugin provides location awareness to your mobile application. Enabling this plugin provides a
+ * contextual experience to your users by showing an icon representing the users current location. A few different modes
+ * are offered to provide the right context to your users at the correct time. {@link MyLocationLayerMode#TRACKING}
+ * simply shows the users location on the map represented as a dot. {@link MyLocationLayerMode#COMPASS} mode allows you
+ * to display an arrow icon (by default) that points in the direction the device is pointing in.
+ * {@link MyLocationLayerMode#NAVIGATION} can be used in conjunction with our Navigation SDK to display a larger icon we
+ * call the user puck. Lastly, {@link MyLocationLayerMode#NONE} can be used to disable the Location Layer but keep the
+ * instance around till the activity is destroyed.
+ * <p>
+ * Using this plugin requires you to request permission beforehand manually or using
+ * {@link com.mapbox.services.android.telemetry.permissions.PermissionsManager}. Either {@code ACCESS_COARSE_LOCATION}
+ * or {@code ACCESS_FINE_LOCATION} permissions can be requested and this plugin work as expected.
+ *
+ * @since 0.1.0
+ */
+@SuppressWarnings( {"MissingPermission"})
+public class MyLocationLayerPlugin implements LocationEngineListener, CompassListener {
```

Comment:
This class could be decoupled into more smaller focused classes, as a result this would shrink most methods to a couple of lines.